### ISSUE TITLE
Allow concat_map to accept a string or integer

### ISSUE THREAD

**Issue by deontologician** [rethinkdb/rethinkdb#4459]
Implicitly, `.concat_map(string_or_int)` would translate to `.concat_map(lambda row: row[string_or_int])`

**Comment by marshall007** [comment]
+1 I almost always try to do this before remembering that you can't. Should this also apply to `.map()` and maybe even `.filter()`?

**Comment by deontologician** [comment]
@mlucy mentioned there was an issue for map to accept a string, but I can't find it. Seems like it would be useful in filter too

**Comment by danielmewes** [comment]
What would it do in filter? Check if the field exists? Or if the value of the field is true?

**Comment by deontologician** [comment]
It could plausibly do either. I'd say getting the value of the field would be the more obvious interpretation, but probably the less useful one

**Comment by danielmewes** [comment]
> It could plausibly do either.

I think that might be an indication that we should not do it for `filter`. Especially since the difference can be pretty subtle if we use truthyness.

**Comment by deontologician** [comment]
Yeah, I don't have a strong opinion on it, you're probably right that it's ambiguous enough not to be worth the minor convenience

**Comment by mlucy** [comment]
There's an issue somewhere for making `r.range(10).map(1)` identical to `r([1])*10`.

I think we should just do this for `concat_map` for now.

**Comment by AtnNn** [comment]
`fitler` already converts its arguments to functions. For comparison, here's a partial survey of the current implicit conversions that RethinkDB does:

Convert to a pagination function: `http`
Convert to a constant function: `merge`, `filter`, `contains`\*, `replace`, `for_each`
Convert to a bracket function: `sum`, `avg`, `min`, `max`, `group`, `order_by`

(\*) Contains treats functions and atoms differently if there are multiple arguments.

**Comment by mlucy** [comment]
@AtnNn -- I wouldn't say `contains` converts to a constant function.  `r.table('test').contains(C)` is the same as `r.table('test').contains(lambda x: x.eq(C))`, not `r.table('test').contains(lambda x: C)`.

**Comment by AtnNn** [comment]
@mlucy thanks for pointing that out.

**Comment by marshall007** [comment]
> @danielmewes: What would it do in filter? Check if the field exists? Or if the value of the field is true?

I was thinking it would just convert the string/int argument to a bracket term rather than doing anything special. Just like what's being proposed for `concat_map`.

**Comment by danielmewes** [comment]
@marshall007 So it would check if the value of respective field is truthy? I still think this is a little bit ambiguous, so I'd prefer not performing this conversion for `filter`.

**Comment by marshall007** [comment]
@danielmewes the current behavior is arguably more confusing than a truthy check would be. Since it converts to a constant function `.filter('field')` is effectively `.filter(true)` and this is very unintuitive.

My main point is that `.filter(r.row('field'))` and `.filter('field')` should produce the same results.

**Comment by danielmewes** [comment]
@marshall007 
> the current behavior is arguably more confusing than a truthy check would be. Since it converts to a constant function [...]

Ah, I forgot we were doing that. I absolutely agree that the current version is already bad.
This adds a third possible interpretation to the mix though, which doesn't make it better. I think we should remove the current implicit conversion to a constant function with `filter`.

**Comment by marshall007** [comment]
@danielmewes perhaps, but given that `filter` already does truthy checks I think `.filter(r.row('field')) === .filter('field')` is the only logically consistent interpretation.

I think where we've gotten into trouble (i.e. what causes ambiguity) are the terms where this is not the case. Another example would be `.count('field')` differing from `.filter(r.row('field')).count()`. In all such cases it's difficult to figure out what's going on under the hood.

So whenever possible, I think you should be able to assume a string argument will be converted to bracket term.

**Comment by mlucy** [comment]
It's not obvious to me that `filter` and `concatMap` should use the same shortcut rule (in general we use different shortcut rules for different commands, @AtnNn enumerates them above).  I think we should discuss which shortcut rule to use for `concatMap` in this issue and which shortcut rule to use for `filter` in #4480 (which @deontologician just opened).

For the narrowed scope of `concatMap` I think we should make `foo.concat_map(x)` identical to `foo.concat_map{|row| row[x]}`.
